libraries/chain/include/eosio/chain/host_context.hpp

```diff
@@ -0,0 +1,601 @@
+
+   int  db_store_i64( name scope, name table, const account_name& payer, uint64_t id, const char* buffer, size_t buffer_size );
+   void db_update_i64( int iterator, account_name payer, const char* buffer, size_t buffer_size );
+   void db_remove_i64( int iterator );
+   int  db_get_i64( int iterator, char* buffer, size_t buffer_size );
+   int  db_next_i64( int iterator, uint64_t& primary );
+   int  db_previous_i64( int iterator, uint64_t& primary );
+   int  db_find_i64( name code, name scope, name table, uint64_t id );
+   int  db_lowerbound_i64( name code, name scope, name table, uint64_t id );
+   int  db_upperbound_i64( name code, name scope, name table, uint64_t id );
+   int  db_end_i64( name code, name scope, name table );
+
+private:
+
+   const table_id_object* find_table( name code, name scope, name table );
+   const table_id_object& find_or_create_table( name code, name scope, name table, const account_name &payer );
+   void                   remove_table( const table_id_object& tid );
+   int  db_store_i64( name code, name scope, name table, const account_name& payer, uint64_t id, const char* buffer, size_t buffer_size );
+
+   /// Action methods
+public:
+   vector<account_name> get_active_producers() const;
+
+   virtual int get_action( uint32_t type, uint32_t index, char* buffer, size_t buffer_size )const = 0;
+   virtual int get_context_free_data( uint32_t index, char* buffer, size_t buffer_size )const = 0;
+   virtual bool is_context_free()const = 0;
+   virtual bool is_privileged()const = 0;
+   action_name get_receiver()const { return receiver; };
+   virtual const action& get_action()const = 0;
+   virtual action_name get_sender() const = 0;
+
+   /// Execution methods:
+
+   // sync calls can be initiated from actions or other sync calls
+   uint32_t execute_sync_call(name receiver, uint64_t flags, std::span<const char> data);
+   uint32_t get_call_return_value(std::span<char> memory) const;
+
+   virtual uint32_t get_call_data(std::span<char> memory) const = 0;
+   virtual void set_call_return_value(std::span<const char> return_value) = 0;
+
+   virtual void execute_inline( action&& a ) = 0;
+   virtual void execute_context_free_inline( action&& a ) = 0;
+   virtual void schedule_deferred_transaction( const uint128_t& sender_id, account_name payer, transaction&& trx, bool replace_existing ) = 0;
+   virtual bool cancel_deferred_transaction( const uint128_t& sender_id ) = 0;
+
+   /// Fields:
+public:
+
+   controller&                      control;
+   chainbase::database&             db;  ///< database where state is stored
+   transaction_context&             trx_context; ///< transaction context in which the action is running
+   account_name                     receiver; ///< the code that is currently running
+   std::vector<char>                action_return_value;
+   std::optional<std::vector<char>> last_sync_call_return_value{}; /// return value of last sync call initiated by the current code (host context)
```

Comment:
I should have thought about this in a prior PR, but afaict there doesn't seem to be any need to make this an optional?

Reply:
It does not need to be optional as we only return the size of the return value. 0 can be for both 0-length return value or return value not existing. Will make the change after I resolve the large number of files of merge conflicts.

Reply:
Done. https://github.com/AntelopeIO/spring/pull/1273/commits/6a079409221c7290f6c4dee8f92e9fff4c7cb0f5